File: go/src/step0_repl/step0_repl.go
package main

import (
	"fmt"
	"strings"

	"readline"
)

func Read(raw string) string {
	return raw
}

func Eval(read string) string {
	return read
}

func Print(value string) string {
	return value
}

func rep(input string) string {
	return Print(Eval(Read(input)))
}

func main() {
	for {
		line, err := readline.Readline("user> ")
		if err != nil {
			break
		}
		line = strings.TrimRight(line, "\n")
		fmt.Println(rep(line))
	}
}

File: impls/cljvm/src/mal/steps/step0_repl.clj
(ns mal.steps.step0_repl
  (:require
   [mal.util.malli :as mu]))

(mu/defn mal-read :- :string
  [input :- :string]
  input)

(mu/defn mal-eval :- :string
  [ast :- :string]
  ast)

(mu/defn mal-print :- :string
  [value :- :string]
  value)

(mu/defn rep :- :string
  [input :- :string]
  (-> input
      mal-read
      mal-eval
      mal-print))

(defn -main []
  (loop []
    (print "user> ")
    (flush)
    (let [input (read-line)]
      (if (and input (pos? (count input)))
        (do
          (println (rep input))
          (recur))
        :eof))))
